### ISSUE TITLE
Inconsistent implicit conversion between types in debug and release builds when export hint doesn't match default initialization

### ISSUE THREAD

**Issue by TomWor** [godotengine/godot#39215]
**Godot version:**
3.2 Stable

**Issue description:**
This code doesn't lead to an editor warning (exporting int when value is String).
It does run in the editor when pressing play, but it does not work on export.
There it gives the error message: "Invalid operands 'String' and 'String' in operator '-(negation)'"

```
export(int) var speed = "8"
...
func _moveTrack(delta):
    for tile in current_active_track_tiles:
        tile.translate(Vector3(0,0,-speed * delta))
```

**Comment by JulianDH97** [comment]
It should work if you remove the quotation marks. 
`export(int) var speed = 8`

**Comment by ThakeeNathees** [comment]
when you press play press play, the editor casting the variable speed to `int` by the export hint (int). but with out the editor it'll be `"8"` string and that's why the error, I don't think it's a bug, but need core developers' opinion for the expected behavior.
and adding types everywhere gives warning like `export(int)  var x:String = '8'`.

**Comment by TomWor** [comment]
@JulianDH97 Thanks, I had already figured that out, just wanted to file it here as I think this is a bug.

@ThakeeNathees I'll use static types more often now, it seems a good idea to catch those issues.
Your explanation makes sense but I still thing this should be considered an issue. It should at least print a warning in the console. I tested the game in-editor and everything worked fine until I exported the project and then the issue popped up - I think this should at least be seen as a usability issue that could be smoothed out by a warning or error in the console.

**Comment by hilfazer** [comment]
> Invalid operands 'String' and 'String' in operator '-(negation)'
Weird message. 2 operands for an unary operator?

Otherwise there's no bug - your variable was typed dynamically so it was possible to assign anything to it. Export hint only limits what can be set via Inspector.

**Comment by TomWor** [comment]
Shouldn't the editor play action closely match the behaviour of the exported binary?

**Comment by Calinou** [comment]
> Shouldn't the editor play action closely match the behaviour of the exported binary?

In theory, yes. In practice, subtle bugs like these can appear as the exported binary doesn't have RTTI (run-time type information) to decrease its binary size.

**Comment by akien-mga** [comment]
Confirmed in `3.x` (7244ea7e20bdb5d0aa5fde61a82f958b1bc5bbec), it's probably a duplicate of #56609.

**Comment by vnen** [comment]
I realized I didn't comment on this, so here's my thoughts on this.

> * Don't coerce variable types to that set in the export hint on debug. So `export(int) var speed = "8"` would still be a String and behave a such, unless `speed` is changed in the Inspector, which will present an int spin slider. That may lead to surprises for the developer, but `export(int) var speed = "8"` is invalid code and asking for trouble :)

I do agree not coercing is the way to go, but doing so will reintroduce #27575 (which was fixed by adding this coercion). I couldn't figure out yet another for that. That's the reason I haven't done a PR for this yet.

> * To mitigate the above, a warning should be introduced for when the export hint type doesn't match the type of the default value. This could likely also be raised when the export hint type (`export(int)`) doesn't match an actual type hint (inferred or explicit). So all these cases should IMO raise a warning:

We could add the warning but keep the coercion for now, since the above bug is arguably worse than this one. At least until we can find a proper fix for that. The warning would tell that those values are coerced only in editor, not when exported.

**Comment by akien-mga** [comment]
Fixed by #58686.
